Mathlib/Algebra/Algebra/Bilinear.lean

```diff
@@ -273,3 +273,23 @@ theorem mul_injective [NoZeroDivisors A] {x : A} (hx : x ≠ 0) : Function.Injec
 end Ring
 
 end LinearMap
+
+/-- If `E` is an `F`-algebra which is also a free `F`-module of rank one,
+then the algebra map from `F` to `E` is an isomorphism. -/
+theorem bijective_algebraMap_of_linearEquiv
+    {F E : Type*} [CommRing F] [Ring E] [Algebra F E]
+    (b : F ≃ₗ[F] E) : Function.Bijective (algebraMap F E) := by
+  have h : b.symm (b 1 * b (b.symm 1)) = b.symm 1 * b.symm (b 1 * b 1) := by
+    let j := b.symm.toLinearMap ∘ₗ .mulLeft F (b 1) ∘ₗ b.toLinearMap
+    change j (b.symm 1) = (b.symm 1) • j 1
+    rw [← map_smul, smul_eq_mul, mul_one]
+  replace h : b 1 = (algebraMap F E) (b.symm (b 1 * b 1)) := by
+    apply_fun b at h
+    rwa [b.apply_symm_apply, b.apply_symm_apply, mul_one, mul_comm (b.symm 1),
+      ← smul_eq_mul, map_smul, b.apply_symm_apply, Algebra.smul_def, mul_one] at h
+  refine Function.bijective_iff_has_inverse.2 ⟨fun y ↦ b.symm y * b.symm (b 1 * b 1),
+    fun x ↦ b.injective ?_, fun y ↦ ?_⟩
+  · dsimp only; rw [mul_comm, ← smul_eq_mul, map_smul, Algebra.smul_def, ← h, b.apply_symm_apply,
+      ← Algebra.commutes x (b 1), ← Algebra.smul_def, ← map_smul, smul_eq_mul, mul_one]
+  · dsimp only; rw [_root_.map_mul, ← h, ← Algebra.smul_def, ← map_smul, smul_eq_mul, mul_one,
+      b.apply_symm_apply]
```

Comment:
I don't think this belongs here. Either

```
/-- If `E` is an `F`-algebra which is also a free `F`-module of rank one,
then the algebra map from `F` to `E` is an isomorphism. -/
theorem bijective_algebraMap_of_linearEquiv
    {F E : Type*} [CommSemiring F] [Semiring E] [Algebra F E]
    (b : F ≃ₗ[F] E) : Function.Bijective (algebraMap F E) := by
  have h : b 1 = (algebraMap F E) (b.symm (b 1 * b 1)) := by
    apply b.symm.injective
    rw [Algebra.algebraMap_eq_smul_one, map_smul, b.symm_apply_apply, smul_eq_mul,
      mul_comm, ← smul_eq_mul, ← map_smul, ← smul_mul_assoc, ← map_smul, smul_eq_mul, mul_one,
      b.apply_symm_apply, one_mul, b.symm_apply_apply]
  refine Function.bijective_iff_has_inverse.2 ⟨fun y ↦ b.symm y * b.symm (b 1 * b 1),
    fun x ↦ b.injective ?_, fun y ↦ ?_⟩
  · dsimp only; rw [mul_comm, ← smul_eq_mul, map_smul, Algebra.smul_def, ← h, b.apply_symm_apply,
      ← Algebra.commutes x (b 1), ← Algebra.smul_def, ← map_smul, smul_eq_mul, mul_one]
  · dsimp only; rw [_root_.map_mul, ← h, ← Algebra.smul_def, ← map_smul, smul_eq_mul, mul_one,
      b.apply_symm_apply]
```
That only uses the definition of `Algebra` and `LinearEquiv`, or even 
```

variable {F E : Type*} [CommSemiring F] [Semiring E] [Algebra F E] (b : F →ₗ[F] E)

lemma injective_algebraMap_of_linearMap (hb : Function.Injective b) :
    Function.Injective (algebraMap F E) := fun x y e ↦ hb <| by
  rw [← mul_one x, ← mul_one y, ← smul_eq_mul, ← smul_eq_mul,
      map_smul, map_smul, Algebra.smul_def, Algebra.smul_def, e]

lemma surjective_algebraMap_of_linearMap (hb : Function.Surjective b) :
    Function.Surjective (algebraMap F E) := by
  intro x
  obtain ⟨x, rfl⟩ := hb x
  obtain ⟨y, hy⟩ := hb (b 1 * b 1)
  refine ⟨x * y, ?_⟩
  obtain ⟨z, hz⟩ := hb 1
  apply_fun (x • z • ·) at hy
  rwa [← map_smul, smul_eq_mul, mul_comm, ← smul_mul_assoc, ← map_smul _ z, smul_eq_mul, mul_one,
    ← smul_eq_mul, map_smul, hz, one_mul, ← map_smul, smul_eq_mul, mul_one, smul_smul,
      ← Algebra.algebraMap_eq_smul_one] at hy

lemma bijective_algebraMap_of_linearMap (hb : Function.Bijective b) :
    Function.Bijective (algebraMap F E) :=
  ⟨injective_algebraMap_of_linearMap b hb.1, surjective_algebraMap_of_linearMap b hb.2⟩
```
That can go somewhere even earlier.

Reply:
Let me see.

Reply:
`#find_home` suggests `Mathlib.Algebra.Algebra.Defs`. But is it a good idea to put them into such a definition file?
Or it's better to put them into `Mathlib.Algebra.Algebra.Basic` where `LinearEquiv` is imported.

Reply:
Can you also fix the docstring? The actual statement does not mention free modules.

Reply:
The latter does not need `LinearEquiv`. I was thinking maybe `Mathlib/Algebra/Algebra/Basic.lean`.

Reply:
> Can you also fix the docstring? The actual statement does not mention free modules.

Sure, will do later.